code/src/OSFramework/Grid/AbstractDataSource.ts

```diff
@@ -0,0 +1,222 @@
+            });
+        });
+    }
+
+    export abstract class AbstractDataSource implements IDataSource {
+        protected _convertions: Map<string, Set<string>>;
+        // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/no-explicit-any
+        protected _ds: Array<any>;
+        protected _metadata: JSON;
+
+        constructor() {
+            // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/no-explicit-any
+            this._ds = new Array<any>();
+            this._convertions = new Map<string, Set<string>>();
+        }
+
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        protected _getChangesString(itemsChanged: any): string {
+            let tempArray = itemsChanged.map((p) => _.cloneDeep(p));
+
+            //In-place convert data to Outsystems Format
+            ToOSFormat(this._convertions, tempArray);
+
+            if (this.isSingleEntity) {
+                //if the line has a single entity or structure, let's flatten it, so that we avoid the developer
+                //when deserializing to need to put in the JSONDeserialize in the target "List Record {ENTITY}" -> would require extra step.
+                tempArray = FlattenArray(tempArray);
+            }
+
+            return JSON.stringify(tempArray);
+        }
+
+        /**
+         * Parse JSON and get the structure of the new item.
+         * @param json
+         */
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        protected _parseNewItem(): any {
+            const parsedNewItem = _.cloneDeep(this._metadata);
+            const converter = (object) => {
+                Object.keys(object).forEach((key) => {
+                    if (typeof object[key] === 'object') converter(object[key]);
+                    else object[key] = undefined;
+                });
+            };
+
+            converter(parsedNewItem);
+
+            return parsedNewItem;
+        }
+
+        public get hasMetadata(): boolean {
+            return !!this._metadata;
+        }
+
+        public get isSingleEntity(): boolean {
+            return Object.keys(this._ds[0] || {}).length <= 1;
+        }
+
+        public addRow(position?: number, data?: JSON[]) {
+            this._ds.splice(position, 0, ...data);
+        }
+
+        public flatten(): void {
+            const newData = FlattenArray(this._ds);
+            this._ds.splice(0);
+            this._ds.push(...newData);
+        }
+
+        public getData(): JSON[] {
+            return this._ds;
+        }
+
+        public getMetadata(): JSON {
+            return this._metadata;
+        }
+
+        public removeRow(item: number | JSON): boolean {
+            const index =
+                typeof item === 'number' ? item : this._ds.indexOf(item);
+
+            if (index === -1) return false;
+
+            this._ds.splice(index, 1);
```

Comment:
What happens when the index is out of boundaries?
Maybe we should protect this code a bit more.

Reply:
const months1 = ['Jan', 'March', 'April', 'June'];
months1.splice(1, 1, 'May');
console.log(months1);
// Array ["Jan", "Maasdfasdy", "April", "June"]

const months2 = ['Jan', 'March', 'April', 'June'];
months2.splice(123523512, 1, 'May');
console.log(months2);
// Array ["Jan", "March", "April", "June", "May"]

Reply:
It seems ok to me. @rugoncalves

Reply:
perfect